src/pydap/handlers/dap.py

```diff
@@ -85,41 +86,60 @@ def __init__(
         self.verify = verify
         self.checksums = checksums
         self.user_charset = user_charset
-        self.url = url
-
-        # urlparse returns an additional var compared to
-        # urlsplit: `param`. Will toss it.
-        scheme, netloc, path, _, query, fragment = urlparse(self.url)
-        self.scheme = scheme
-        self.netloc = netloc
-        self.path = path
-        self.query = query
-        self.fragment = fragment
-
-        if protocol:
-            if protocol not in ["dap2", "dap4"]:
-                raise TypeError("protocol must be one of `dap2` or `dap4")
-            self.protocol = protocol
-            if self.scheme == protocol:
-                # the other alternative occurs during testing
-                # the server - only when protocol and scheme match,
-                # should pydap change the scheme provided by user
-                self.scheme = "https"
-        else:
-            self.protocol = self.determine_protocol()
         self.get_kwargs = get_kwargs or {}
 
-        self.projection, self.selection = parse_ce(self.query, self.protocol)
-        arg = (
-            self.scheme,
-            self.netloc,
-            self.path,
-            "",
-            "&".join(self.selection),
-            self.fragment,
-        )
-        self.base_url = urlunparse(arg)
-        self.make_dataset()
+        if not url.startswith("dap4") and url.endswith(".dmrpp"):
+            self.protocol = "dap4"
+            if url.startswith("http://") or url.startswith("https://"):
+                r = session.get(url, stream=True)
+                dmrpp = io.BytesIO(r.content.decode())
+                dmrpp_instance = DMRPPPparser(root=ET.parse(dmrpp).getroot())
+            else:
+                dmrpp = open(url).read()
+                dmrpp_instance = DMRPPPparser(root=ET.fromstring(dmrpp))
+            self.projection = []
+            self.dataset = dmrpp_instance.to_dataset()
+            self.base_url = dmrpp_instance.data_filepath
+            if not self.base_url:
+                # some dmrpps do not have the opendap url embedded in their
+                if url.endswith(".dap.dmrpp"):
+                    self.base_url = url.removesuffix(".dap.dmrpp")
+                else:
+                    self.base_url = url.removesuffix(".dmrpp")
+        else:
```

Comment:
One thing to know is that Hyrax in NASA's EDC supports DMR++ retrieval using the suffix '.dap.dmrpp'. That should return the DMR++ XML document and I think it performs the URL template resolution, too. Yes. Here's an example: https://opendap.earthdata.nasa.gov/collections/C1996546500-GHRC_DAAC/granules/f16_ssmis_20250221v7.nc.dap.dmrpp

Reply:
About the URL template resolution, you mean for example the: 
```python
xml:base="http://test.opendap.org/opendap/dap4/TestGroupData.nc4.h5"
```
element in the dmrpp? I have seen some `dmrpp`s without this `xml:base` element. And so this conditional is a last resource for finding the opendap url to the file. With earthaccess + virtualizarr, this filepath is provided, but this enables a few more cases (when it is not provided).

Reply:
OK. I remember we had some issues with the xml:base URL in the past. But for the 'Template URL' I meant the data URL the replaces the template code that appears in the XML Dataset element's 'href' attribute. When the documents are written by the code that ingests data in the NASA cloud, the URL to the data granule is not set as te value of that attribute. Instead there is some string like "OPeNDAP_Data_URL" or something like that and the server performs a text substitution before the data access code parses and interprets the DMR++.